### ISSUE TITLE
Update JwtBearerOptions via DI after configuration when secret or issuer changes

### ISSUE THREAD

**Issue by terryaney** [dotnet/aspnetcore#49586]
Similar to [Better way to get the Json Web Key in AddJwtBearer](https://github.com/dotnet/aspnetcore/issues/21491) and [Use object in DI when AddJwtBearer](https://github.com/aspnet/Security/issues/1709), I wanted a way to have my Jwt secret/issuer come from a configuration source that could be updated and have our sites/apis resilient enough to automatically use them if they changed without requiring a restart of each site that depended on the credentials.

I wanted to follow up with this question on either of those tasks but both were closed.  I don't know if @HaoK wants to weigh in a bit since his 'tip' is what got Voccano's code working in issue #21491, but the issue is, for some reason DI for `TheKeep` (my object) was successfully injected, but the exposed properties were not updated at the time of my `UpdatableJwtBearerOptions` class.  When it was ultimately injected into the minimal api endpoint call, they WERE updated there.

The relevant code is below, and I'll post screen shots after that represent:
1. Start api with issuer set to 'bad.*'
2. Make a request with 'bad.*' and receive 401 unauthorized (correct).
3. Update configuration source to remove 'bad.'
4. Make a request and receive 200 (correct)
5. First screen will be debug inside `UpdatableJwtBearerOptions` and second will be inside minimal api endpoint.

Anything obvious as to why `TheKeep` wouldn't work?

Configure JwtBearerOptions and 'TheKeep' (my configuration class)
```csharp
builder.Services
	.Configure<TheKeepSettings>( builder.Configuration.GetSection( "TheKeep" ) ) // configuration in here
	.AddScoped<Domain.Configuration.TheKeep>()
builder.Services.AddTransient<IConfigureOptions<JwtBearerOptions>, UpdatableJwtBearerOptions>();
```

TheKeep snippet showing DI of 'IConfiguration' and Jwt property
```csharp
public TheKeep( IOptionsSnapshot<TheKeepSettings> theKeepSettings )
{
	this.Jwt = theKeepSettings.Value.Jwt;
}
public Jwt Jwt { get; init; }
```

UpdatableJwtBearerOptions Class implementation to set Jwt secret/issuer
```csharp
public class UpdatableJwtBearerOptions : IConfigureNamedOptions<JwtBearerOptions>
{
	private readonly IConfiguration configuration;
	private readonly Domain.Configuration.TheKeep theKeep;

	public UpdatableJwtBearerOptions( IConfiguration configuration, Domain.Configuration.TheKeep theKeep )
	{
		this.configuration = configuration;
		this.theKeep = theKeep;
	}

	public void Configure( string? name, JwtBearerOptions options )
	{
		options.TokenValidationParameters.IssuerSigningKeyResolver = ( token, securityToken, kid, validationParameters ) =>
		{
			validationParameters.ValidIssuer = configuration.GetValue<string>( "TheKeep:Jwt:WebServiceProxy:Issuer" );
			var issuerFromTheKeep = theKeep.Jwt.WebServiceProxy.Issuer;

			return new List<SecurityKey>() 
			{ 
				new SymmetricSecurityKey( Encoding.ASCII.GetBytes( configuration.GetValue<string>( "TheKeep:Jwt:WebServiceProxy:Secret" )! ) )
			};
		};
	}

	public void Configure( JwtBearerOptions options ) => throw new NotImplementedException();
}
```

Snippet of minimal api...
```csharp
public override async Task HandleAsync( NexgenRequest r, CancellationToken c )
{
	var issuerFromTheKeep = theKeep.Jwt.WebServiceProxy.Issuer;
	// ... code omitted
}
```

`TheKeep` not updated yet...
![image](https://github.com/dotnet/aspnetcore/assets/6237247/929787af-d1d7-4324-b17e-27f634760cc4)

`TheKeep` updated...
![image](https://github.com/dotnet/aspnetcore/assets/6237247/9d0a47ac-d598-4f43-9961-120880efdbd9)

**Comment by javiercn** [comment]
@terryaney thanks for contacting us.

Are you using hot reload? Can you turn it off and see if it reproduces?

**Comment by terryaney** [comment]
> @terryaney thanks for contacting us.
> 
> Are you using hot reload? Can you turn it off and see if it reproduces?

Embarrassed....I don't think I am using hot reload, but can you let me know how I can confirm?

**Comment by mkArtakMSFT** [comment]
The answer in [this post](https://stackoverflow.com/questions/68952259/disabling-hot-reload-for-net-core-project-in-visual-studio-2019) will help. Try to disable it and see if the behavior changes.

**Comment by terryaney** [comment]
I'm using VS Code.  I don't think I have hot reload enabled.

**Comment by terryaney** [comment]
> @terryaney where is the configuration stored? Is it `appSettings.json`?
> 
> Also, if you haven't come across the docs for monitoring configuration changes, this may help too: https://learn.microsoft.com/aspnet/core/fundamentals/change-tokens?view=aspnetcore-7.0#monitor-for-configuration-changes

@mkArtakMSFT It is stored in a json file outside of the appsettings.  The json file is added to the configuration via:

`builder.Configuration.AddJsonFile( Path.Combine( csPath, "Camelot.Secrets.json" ), optional: true, reloadOnChange: true )`

The problem isn't about the update being triggered, b/c as I said, it works with an `IConfiguration` object passed in, the problem is simply that 'theKeep' (which itself uses an `IOptionsSnapshot` via DI) is not updated.

**Comment by terryaney** [comment]
@halter73 Thanks for the ideas.  So I didn't provide enough code I don't think :)  But maybe you already knew what I was doing.

**Here is my builder registration**
```csharp
builder.Services.AddTransient<IConfigureOptions<JwtBearerOptions>>( provider =>
{
	var configuration = provider.GetRequiredService<IConfiguration>();
	using( var scope = provider.CreateScope() )
	{
		var theKeep = scope.ServiceProvider.GetRequiredService<TheKeep>();
		var options = new UpdatableJwtBearerOptions( configuration, theKeep );
		return options;
	}
});
```
And from above, my constructor for `UpdatableJwtBearerOptions` is:

```csharp
public UpdatableJwtBearerOptions( IConfiguration configuration, Domain.Configuration.TheKeep theKeep )
{
	this.configuration = configuration;
	this.theKeep = theKeep;
}
```

As is, you were right.

1.  I only hit `Configure` one time, but hit the `IssuerSigningKeyResolver` delegate every request.
2.  Inside the delegate, if I modified my *.json file, the `configuration.GetValue<string>()` was updated, but theKeep value wasn't.
3. I put a breakpoint in the ctor for `UpdatableJwtBearerOptions` and as you probably know, it was only hit once, which surprised me since it was registered as *Transient*.  So obviously, my reference to `theKeep` was never updated.

I put in your `builder.Services.AddSingleton<IOptionsChangeTokenSource<JwtBearerOptions>>` suggestion and:

1. The Configure method was triggered as soon as there was an edit to the file which simply re-assigned the `IssuerSigningKeyResolver` delegate.
2. The ctor was *not* called, so I was stuck with the old `theKeep` reference.

So currently, I simply work with an `IConfiguration` object inside `UpdatableJwtBearerOptions`, don't need to resolve a *scoped* `TheKeep` to pass in, and not using your 'TokenSource' suggestion.  The only downside to this is accessing the configuration via `GetValue<string>( path )` instead of strongly typed properties of `TheKeep`.

If you see anything that might allow me to use `TheKeep`, let me know.
